File: api/app/services/auth_service.py
"""
Authentication service for user management and JWT operations.
"""


from sqlalchemy.orm import Session

from app.models.organization import MemberRole
from app.models.user import User
from app.repositories.organization_repository import OrganizationMemberRepository
from app.repositories.user_repository import UserRepository
from app.schemas.auth import Token, UserCreate, UserLogin
from app.services.base import BaseService
from app.services.sms_service import SmsService, get_sms_provider
from app.utils.security import (
    create_access_token,
    create_refresh_token,
    decode_token,
    get_password_hash,
    verify_password,
)


class AuthService(BaseService):
    """Service for authentication and user management."""

    def __init__(self, db: Session):
        super().__init__(db)
        self.user_repo = UserRepository(db)
        self.member_repo = OrganizationMemberRepository(db)
        self.sms_service = SmsService(db, get_sms_provider())

    def register(self, user_data: UserCreate) -> User:
        """
        Register a new user and create a personal team.

        Args:
            user_data: User registration data

        Returns:
            Created user

        Raises:
            ValueError: If phone already exists or verification code is invalid
        """
        # 1. 验证验证码
        if not self.sms_service.verify_code(user_data.phone, user_data.verification_code, "register"):
            raise ValueError("验证码无效或已过期")

        # 2. 检查手机号是否已注册
        if self.user_repo.exists_by_phone(user_data.phone):
            raise ValueError("手机号已注册")

        # 3. 创建用户
        user = User(
            phone=user_data.phone,
            password_hash=get_password_hash(user_data.password),
            full_name=user_data.full_name,
        )
        user = self.user_repo.create(user)

        # 4. 创建个人团队
        from app.services.organization_service import OrganizationService

        org_service = OrganizationService(self.db)
        org_service.create_personal_team(
            user_id=user.id,
            user_name=user.full_name or "用户",
        )

        return user

    def login(self, credentials: UserLogin) -> Token:
        """
        Authenticate user and return tokens.
        Supports password or verification code login.

        Args:
            credentials: Login credentials (phone + password or phone + verification_code)

        Returns:
            Token with access and refresh tokens

        Raises:
            ValueError: If credentials are invalid
        """
        user = self.user_repo.find_by_phone(credentials.phone)
        if not user:
            raise ValueError("手机号或密码错误")

        if credentials.password:
            # 密码登录
            if not verify_password(credentials.password, user.password_hash):
                raise ValueError("手机号或密码错误")
        else:
            # 验证码登录
            if not self.sms_service.verify_code(credentials.phone, credentials.verification_code, "login"):
                raise ValueError("验证码无效或已过期")

        token_data = {"sub": str(user.id), "phone": user.phone}
        access_token = create_access_token(token_data)
        refresh_token = create_refresh_token(token_data)

        return Token(
            access_token=access_token,
            refresh_token=refresh_token,
            token_type="bearer",
        )

    def send_login_code(self, phone: str) -> bool:
        """
        Send login verification code.

        Args:
            phone: Phone number

        Returns:
            bool: Whether code was sent successfully
        """
        return self.sms_service.send_code(phone, "login")

    def send_register_code(self, phone: str) -> bool:
        """
        Send registration verification code.

        Args:
            phone: Phone number

        Returns:
            bool: Whether code was sent successfully

        Raises:
            ValueError: If phone already registered
        """
        if self.user_repo.exists_by_phone(phone):
            raise ValueError("手机号已注册")
        return self.sms_service.send_code(phone, "register")

    def refresh_token(self, refresh_token: str) -> Token:
        """
        Refresh access token using refresh token.

        Args:
            refresh_token: Valid refresh token

        Returns:
            New token pair

        Raises:
            ValueError: If refresh token is invalid
        """
        payload = decode_token(refresh_token)
        if not payload or payload.get("type") != "refresh":
            raise ValueError("Invalid refresh token")

        user = self.user_repo.get(payload["sub"])
        if not user:
            raise ValueError("User not found")

        token_data = {"sub": str(user.id), "phone": user.phone}
        new_access_token = create_access_token(token_data)
        new_refresh_token = create_refresh_token(token_data)

        return Token(
            access_token=new_access_token,
            refresh_token=new_refresh_token,
            token_type="bearer",
        )

    def get_current_user(self, user_id: str) -> User | None:
        """Get current user by ID."""
        return self.user_repo.get(user_id)

    def check_permission(self, user_id: str, org_id: str, allowed_roles: list[MemberRole]) -> bool:
        """Check if user has required role in organization."""
        return self.member_repo.has_role(org_id, user_id, allowed_roles)

File: api/src/services/auth.service.ts
import type { AuthRepository } from '../repositories/auth.repo.js';
import { defaultAuthRepo } from '../repositories/auth.repo.js';
import { hashPassword, verifyPassword } from '../utils/security.js';
import { createAccessToken, createRefreshToken, decodeToken } from '../utils/jwt.js';
import { createAppError } from '../utils/appError.js';
import { ulid } from 'ulid';

/**
 * 注册输入（无需短信验证）
 */
export interface RegisterInput {
  phone: string;
  full_name: string;
  password: string;
}

/**
 * 登录输入（仅支持密码登录）
 */
export interface LoginInput {
  phone: string;
  password: string;
}

/**
 * 登录响应
 */
export interface LoginResult {
  access_token: string;
  refresh_token: string;
  token_type: string;
}

/**
 * 用户信息响应
 */
export interface UserInfo {
  id: string;
  phone: string | null;
  full_name: string | null;
  is_active: boolean;
  created_at: Date;
}

/**
 * Auth Service 接口
 */
export interface AuthService {
  register(data: RegisterInput): Promise<UserInfo>;
  login(data: LoginInput): Promise<LoginResult>;
  refreshToken(refreshToken: string): Promise<LoginResult>;
  getUserById(id: string): Promise<UserInfo | null>;
}

/**
 * 创建 Auth Service 实例
 */
export function createAuthService(
  getRepo: () => AuthRepository = () => defaultAuthRepo
): AuthService {
  return {
    register: async (data: RegisterInput) => {
      const { phone, full_name, password } = data;

      // 检查手机号是否已注册
      const existing = await getRepo().findUserByPhone(phone);
      if (existing) {
        throw createAppError(400, '手机号已注册');
      }

      // 创建用户
      const passwordHash = await hashPassword(password);
      const userId = ulid().toLowerCase();
      const user = await getRepo().createUser({
        id: userId,
        phone,
        full_name,
        password_hash: passwordHash,
      });

      return {
        id: user.id,
        phone: user.phone,
        full_name: user.full_name,
        is_active: user.is_active,
        created_at: user.created_at,
      };
    },

    login: async (data: LoginInput) => {
      const { phone, password } = data;

      const user = await getRepo().findUserByPhone(phone);
      if (!user) {
        throw createAppError(401, '手机号或密码错误');
      }

      // 仅支持密码登录
      const ok = await verifyPassword(password, user.password_hash);
      if (!ok) {
        throw createAppError(401, '手机号或密码错误');
      }

      if (!user.is_active) {
        throw createAppError(401, '账号已停用或不允许登录');
      }

      const access_token = createAccessToken({ sub: user.id, phone: user.phone });
      const refresh_token = createRefreshToken({ sub: user.id, phone: user.phone });

      return {
        access_token,
        refresh_token,
        token_type: 'bearer',
      };
    },

    refreshToken: async (refreshToken: string) => {
      const payload = decodeToken(refreshToken);
      if (!payload || (payload.type as string) !== 'refresh') {
        throw createAppError(401, 'Could not validate credentials');
      }

      const sub = payload.sub as string;
      const user = await getRepo().findUserById(sub);
      if (!user || !user.is_active) {
        throw createAppError(401, 'Could not validate credentials');
      }

      const access_token = createAccessToken({ sub: user.id, phone: user.phone });
      const refresh_token = createRefreshToken({ sub: user.id, phone: user.phone });

      return {
        access_token,
        refresh_token,
        token_type: 'bearer',
      };
    },

    getUserById: async (id: string) => {
      const user = await getRepo().findUserById(id);
      if (!user) return null;
      return {
        id: user.id,
        phone: user.phone,
        full_name: user.full_name,
        is_active: user.is_active,
        created_at: user.created_at,
      };
    },
  };
}

/**
 * 默认实例
 */
export const defaultAuthService = createAuthService();
